### ISSUE TITLE
Use of unnecessary fully qualified names in the documentation

### ISSUE THREAD

**Issue by asmeurer** [sympy/sympy#23081]
The documentation is full of functions and classes that are referenced with fully qualified names. For example, at https://docs.sympy.org/dev/modules/sets.html, you'll see the definitions as `sympy.sets.set.Set`, `sympy.sets.set.Interval` and so on. 

This is problematic for any function or class that is present in the top-level `sympy` namespace. We want users to use `sympy.Set`, `sympy.Interval`, and so on. The specific submodule `sympy.sets.set` should be an implementation detail only that users should not be relying on. The extensive use of this in the docs makes it hard to even say that submodules like `sympy.sets.set` *are* an internal detail, even though they really should be. See https://github.com/sympy/sympy/issues/23037.

Additionally, this means that anyone referencing these classes in other Sphinx docstrings will need to reference them by fully qualified name. This is both annoying for the person writing the docstring, who needs to lookup exactly where a given class lives, and also makes it harder to read (for instance, look at the "see also" section for `Set.kind` https://docs.sympy.org/dev/modules/sets.html#sympy.sets.sets.Set.kind, which lists every Set subclass, but in a way that is difficult to read because it uses fully qualified names everywhere. 

This is actually fixable for normal cross references by using the `~.` syntax, e.g., ``` :class:`~.Interval` ```. But this is also annoying to type, and very few people know about it. If you want anything other than that you have to use the general RST link syntax which is even more annoying, e.g., to make `sympy.Interval`, you have to use ``` `sympy.Interval <sympy.sets.set.Interval>`_ ``` (it's even more complicated if you want `sympy.Interval` to be in code text because RST does not allow easily nesting markup). 

I've been investigating if there are ways we can fix this. For RST files where the classes are listed explicitly, you can put

```rst
.. currentmodule:: sympy
```

above any `autoclass` and `autofunction` calls, and it will make them use `sympy.name` instead of the fully-qualified name. For instance, 

```diff
diff --git a/doc/src/modules/sets.rst b/doc/src/modules/sets.rst
index 169c983d44..6b7fdfe9a2 100644
--- a/doc/src/modules/sets.rst
+++ b/doc/src/modules/sets.rst
@@ -9,6 +9,8 @@ Basic Sets

 .. automodule:: sympy.sets.sets

+.. currentmodule:: sympy
+
 Set
 ^^^
 .. autoclass:: Set
```

However, this still has problems:

- It only works for explicitly listed names. For the docs that use

  ```rst
  .. automodule:: sympy.mod
     :members:
  ```

  it does not work. It might be possible to require always listing explicitly all names, but we would have to rely on manually testing documentation coverage using `make coverage` or `bin/coverage_doctest.py` to ensure every function is included in the RST (which to be fair we already ought to be doing anyway).
- It does not completely remove the need to use the `~.` trick for cross references.

Another thing that we can do is to set

```py
add_module_names = False
```

in `conf.py`. This will make it so that the headers for all the functions and classes do not show the module name. However,
- We do want to show the module name in instances where the function/class is not in the top-level namespace (for instance, anything in sympy.physics).
- This does not seem to affect the cross-referencing problem. It only affects how the function/class definition appears.

There is also a `python_use_unqualified_type_names` option, which seems related but I can't figure out if it does anything at all. Whatever it does, it doesn't seem to help with this problem. 

I haven't looked into the autodoc source code yet to see if there isn't some hidden way to fix this that isn't documented in the Sphinx documentation. However, I suspect to really get a good fix for this, we are going to need to write a custom Sphinx extension that hooks into autodoc. Ideally, autodoc would just show the least qualified name it can. So instead of always requiring `sympy.a.b.X`, it would try `sympy.X` and if it isn't in that namespace, it would try `sympy.a.X`, and only if it isn't there would it use the fully-qualified `sympy.a.b.X`. Additionally, I would really like it if simply writing ``` :ref:`X` ``` always first tried `sympy.X` (and maybe also `currentmodule.X`). It would also be nice if it could be smarter about how cross-referencing methods worked, especially for methods that are defined in multiple places.

**Comment by oscargus** [comment]
Isn't there two (different) problems here? I do not really see the problem that the class "heading" has a fully qualified path as it shows the (current) location. It should be more important that the examples do not import with the full path (unless required) and that "See also" (and other links), do not show full path. 

Of course, there is no harm if the full path is not shown, but if that happens to be tricky to avoid (because some do require it), it may be better to focus on the things that probably have a greater impact. (But I may have missed the details here.)

(I've made some efforts regarding imports in the examples and it should be possible to do something similar for "See also", although that is a bit inconsistent as well, sometimes have explanations, e.g. for `Naturals` in the first link you provided, while some are just separated by commas.)

**Comment by asmeurer** [comment]
Doctest examples are a separate thing. I didn't really mention them here because there aren't any technical difficulties in fixing them. I guess we could have some sort of linter that ensures doctests always use non-fully qualified imports when possible. 

The main issue is that as long as we have these fully qualified names in public documentation, we can't really say that the internal layout of things is not public. I see this as sort of a first stepping stone for fixing the broader #23037. For example, how can we really say that some random helper function in `sympy.sets.set` (for example) is private but `sympy.sets.set.ImageSet` is public when they both are documented exactly the same way.

Also, while the header may seem less important, it is still relevant because, at least the way autodoc currently works, it's also the name that you have to use to reference the function. So for example, the `sympy.sets.set.ImageSet` is not just the header (and also the URL anchor, I forgot about that!) that's at https://docs.sympy.org/dev/modules/sets.html#sympy.sets.fancysets.ImageSet, but it's also the text you'll see in most docstrings that cross-reference it. It's only if someone knows about the `~.` trick that you'll see just `ImageSet` (and having `sympy.ImageSet` is impossible unless you use the manual external link syntax). I really want it so that people can just write ``` :class:`ImageSet` ``` or even ``` :ref:`sympy.ImageSet` ```, and it just works. Right now if you do that it complains it doesn't know what `ImageSet` is, so you have to "fix" it by typing out `sympy.sets.set.ImageSet`, even if that's not really what you want the text to say. "See Also" is even worse because it is more limiting in how it lets you spell things (although I think we can adjust the numpydoc extension to improve this).

I agree that having more descriptions on the See Alsos would be great, but that's a separate issue. Actually improving See Alsos would be a nice easy to fix issue.

**Comment by moorepants** [comment]
> we can't really say that the internal layout of things is not public

I'm not sure why we'd want to do that. I think the layout should be public and if people want to import from where something actually is, they should be able to reliably do that.. If we want to be able to move modules around then make them "private" with `_module_name.py`.

**Comment by moorepants** [comment]
I'm of the opinion that we shouldn't be moving modules around that are already publicly available, unless of course we go through a deprecation cycle, but that's still painful for users and should only be done if there is no other choice to solve a technical issues than to move a module.

**Comment by asmeurer** [comment]
Making deprecated stubs just because we want to reorganize the internal structure a bit is annoying. And there's really no reason to do it if we're moving stuff at the top level. 

To me, the docs really should be clear if something is in the top-level. There's no clear distinction there now, unless you look at the doctest examples. Consider for example the logic module. The two functions in [this section](https://docs.sympy.org/dev/modules/logic.html#simplification-and-equivalence-testing) are in the top level. You can just use `sympy.simplify_logic` or `sympy.bool_map`. On the other hand, everything in [this section](https://docs.sympy.org/dev/modules/logic.html#truth-tables-and-related-functions) of the docs is not. If you want to use them you have to import them using `from sympy.logic.boolalg import truth_table`. Just looking at the docs there's no indication of that. 

We could make module names start with an underscore when this happens. That's more of a suggestion for https://github.com/sympy/sympy/issues/23037. I expect it won't be so easy to do that because quite a few files have both top-level (in `__init__.py`) and non-top level functions. So we would have to have underscore files where all the actual code lies and non-underscore stubs for importing. It's not impossible to do. I've seen other Python modules that do this. They all feel rather more difficult to navigate as a contributor than SymPy, however, due to the indirection. 

This issue is just about the Sphinx problems, though. I hope you can agree that even in the cases where you do want a fully qualified name, having to type it out completely just to make a cross-reference work is annoying.

**Comment by moorepants** [comment]
> This issue is just about the Sphinx problems, though. I hope you can agree that even in the cases where you do want a fully qualified name, having to type it out completely just to make a cross-reference work is annoying.

It may be a bit annoying, but I don't think it is a problem. I'm writing a book write now and have to do `:external:py:meth:~sympy.physics.vector.vector.Vector.diff` to link to external sphinx docs. It's a bit long to have to type each time but I haven't found another option. Is seems I have to type the exact path to the primary location of each object. I find my self opening ipthon and using `??` on top level sympy functions to find where exactly they are so I can get this link correct. So, a point is that we do need to see these fully qualified paths if we want to know where the function/class/method actually is in the file directory tree.

For these links, it would be nice if 1) there is a consistent way we type them all, 2) we have to type less, and 3) viewers of the documentation (html and raw text) can still figure out which file contains which object. I'm not sure what the best solution is though.

**Comment by oscarbenjamin** [comment]
> I'm writing a book write now and have to do `:external:py:meth:~sympy.physics.vector.vector.Vector.diff` to link to external sphinx docs. It's a bit long to have to type each time but I haven't found another option.

I think that with Aaron's suggestion here it wouldn't be necessary to do that and you could just give the path to wherever `Vector` is expected to be imported from.

**Comment by smichr** [comment]
> Some authors seem to prefer writing these fully-qualified names

I always use a `path` function to give the path to an object. I don't experiment with trimming the name because I know this works. 

It would be nice to have a public "whereis" function that would find any object specified as a string ('expand') or object (`expand`) that could be imported along with the preferred path and the type of object found (in case there are duplicate).

e.g.
```python
>> from sympy import solve
>> whereis(solve)
`from sympy import solve`
```

Having a way to search docstrings and function/class/method/variable names (perhaps in a given module) would be nice, too. e.g. to find where 'illegal' should be imported from, or `_coeff_isneg`. I often just end up using the texteditor to do this when necessary.

**Comment by smichr** [comment]
> to find where exactly they are so I can get this link correct

This is what I do with the `path` function in my pythonstartups.py file:
```python
import inspect
def source(x):
    print(inspect.getsource(x))

def path(x):
    p = inspect.getsourcefile(x)
    p = p[p.find('sympy')+6:-3]
    n = str(x)
    if n.startswith('<'):
        if '.' in n:
            n = n.split('.')[-1].split("'")[0]
        else:
            n  = n.split()[1]
    return 'from %s import %s' % (p.replace('\\','.'), n)

import pyperclip
copy = lambda x: pyperclip.copy(str(x))
```
copy(path(solve))  # puts "from sympy.solvers.solvers import solve" on the clipboard

**Comment by moorepants** [comment]
I struggled to get any cross referencing to work at all in this recent PR https://github.com/sympy/sympy/pull/23111.

**Comment by moorepants** [comment]
@smichr I just tried your function with `ReferenceFrame.ang_acc_in`:
```python
n [6]: import inspect
   ...: def source(x):
   ...:     print(inspect.getsource(x))
   ...: 
   ...: def path(x):
   ...:     p = inspect.getsourcefile(x)
   ...:     p = p[p.find('sympy')+6:-3]
   ...:     n = str(x)
   ...:     if n.startswith('<'):
   ...:         if '.' in n:
   ...:             n = n.split('.')[-1].split("'")[0]
   ...:         else:
   ...:             n  = n.split()[1]
   ...:     return 'from %s import %s' % (p.replace('\\','.'), n)
   ...: 

In [7]: path(B.ang_acc_in)
Out[7]: 'from physics/vector/frame import ang_acc_in of B>'
```

It should show: `physics.vector.frame.ReferenceFrame.ang_acc_in` to get the class name in there.

**Comment by moorepants** [comment]
> I think that with Aaron's suggestion here it wouldn't be necessary to do that and you could just give the path to wherever Vector is expected to be imported from.

My first attempts at using shorter paths weren't successful, so I just resorted to the full path.

**Comment by asmeurer** [comment]
> My first attempts at using shorter paths weren't successful, so I just resorted to the full path.

Yes, that is the way it is now. The suggestion here is to clean that up, so that you can use the shorter paths and it will actually work. Right now the only thing that works is the fully qualified name.

**Comment by oscargus** [comment]
I stumbled upon https://sphobjinv.readthedocs.io/en/latest/ yesterday. It really helps out when trying to figure out what a "correct" way to reference something is. It may not help with the things discussed here, but made a huge change for me when trying to figure related things out.

**Comment by bskinn** [comment]
Author/maintainer of `sphobjinv` here -- thanks for the plug, @oscargus! 😀

----

@asmeurer:

> Right now the only thing that works is the fully qualified name.

As far as I know, the *only* thing that works is the fully qualified name. 

I have a suggestion for one way to handle this, which I use extensively in my Sphinx docsets. I mostly use it to cut down on the amount of typing I have to do for objects that I want to cross-reference frequently, but I suspect it would also work here to (a) cut down on doc author typing, (b) make it easier for doc authors to write their cross-references correctly, and (c) reinforce the import pattern that you want `sympy` users to follow.

My solution is to use the `rst_epilog` setting in `conf.py`, combined with reST's `replace` functionality. See [here](https://github.com/bskinn/sphobjinv/blob/505c7afc656f20b9e105f4ead9c6eb570eef971a/doc/source/conf.py#L71-L193).  In short, you define in `conf.py` a common epilog/footer that gets appended to every `.rst` file in your documentation, which then allows you to use a common set of `|replace|`s throughout the documentation.

So, taking `sympy.Set` from the original issue post as an example, I used `sphobjinv` to find the right cross-reference:

```
>sphobjinv suggest -u https://docs.sympy.org/latest/index.html class.sympy.sets.set -st80

No inventory at provided URL.
Attempting "https://docs.sympy.org/latest/index.html/objects.inv" ...
Attempting "https://docs.sympy.org/latest/objects.inv" ...
Remote inventory found.

  Name                                              Score
-------------------------------------------------  -------
:py:class:`sympy.sets.sets.Complement`               85
:py:class:`sympy.sets.sets.EmptySet`                 85
:py:class:`sympy.sets.sets.FiniteSet`                85
:py:class:`sympy.sets.sets.Intersection`             85
:py:class:`sympy.sets.sets.Interval`                 85
:py:class:`sympy.sets.sets.ProductSet`               85
:py:class:`sympy.sets.sets.Set`                      85
:py:class:`sympy.sets.sets.SymmetricDifference`      85
:py:class:`sympy.sets.sets.Union`                    85
:py:class:`sympy.sets.sets.UniversalSet`             85
```

Based on this, I might then add an entry like the following to `rst_epilog`:

```
.. |sympy.Set| replace:: :class:`sympy.sets.sets.Set`
```

And then, in a `.rst` doc source, that cross reference could be added just by writing `|sympy.Set|`.

---

This isn't a perfect/universal solution, because the `replace` commands have to be added as they are needed... but the formula for creating new `replace`s in `rst_epilog` is straightforward, and it would be pretty easy to add to a docs contributor's guide. Another friction would come up if the project's module structure were ever refactored, as the `replace`s would all have to be updated. It still seems like it would help a lot.

Separately, I think `rst_epilog`/`replace` is just what you need, @moorepants, for your book.

**Comment by moorepants** [comment]
Nice. I can give it a try!

**Comment by asmeurer** [comment]
I was actually wondering if we could do something similar by adding a cross-reference to each function so that ```` :ref:`Set` ``` would be a cross-reference that points to Set. The only downside would be that you wouldn't be able to use `:func:` or `:class:`, and I'm not sure you could make the automatic adding of parentheses that `:func:` typically does work. If adding a cross-reference to the function itself isn't doable, we could also add them to the headers before the functions (see https://github.com/sympy/sympy/issues/23332).

The replace idea is interesting. It's not ideal because you'd have to use a different cross-reference syntax to make it work. Wouldn't it be possible to automate it by messing with the autodoc templates?

**Comment by bskinn** [comment]
> I'm not sure you could make the automatic adding of parentheses that `:func:` typically does work.

Agreed, I don't think you could make the builtin `:ref:` do that.

> Wouldn't it be possible to automate it by messing with the autodoc templates?

Maybe, but I *think* the challenge with just fiddling with autodoc would be reliably telling autodoc *how* to change what it renders into the built docs (e.g., `sympy.Set` instead of `sympy.sets.sets.Set`).

**Comment by bskinn** [comment]
So, separate from the above... I had a ridiculous idea, but I tested it on one of my own docsets and it seems like it might work.

tldr; point autodoc at the desired import location for each object, instead of the definition location.

Right now, e.g., `sympy.sets.sets.Set` is inserted into the docs using an [`.. autoclass::` directive](https://github.com/sympy/sympy/blob/95d3c22b14d6347efd1bb86e893d80ce568c2a75/doc/src/modules/sets.rst#set). In particular, a directive of the form:

```
.. autoclass:: Set
    :members:
```

What if you changed this to point to the explicit location of where you want people to import it from?

```
.. autoclass:: sympy.Set
    :members:
```

I tried this on an object in one of my own docsets (see [commit](https://github.com/bskinn/sphobjinv/commit/0920e1d410417ca4c1f759db35d6caddcb636c37) on demo branch), and both the definition location (`sphobjinv.inventory.Inventory`) *and* the alternate top-level import location (`sphobjinv.Inventory`) show up in the `objects.inv`:

```
$ sphobjinv su build/html/objects.inv -st60 inventory
 
  Name                                                            Score 
---------------------------------------------------------------  -------
:py:class:`sphobjinv.Inventory`                                    90   <=====
:py:module:`sphobjinv.inventory`                                   90   
:py:class:`sphobjinv.inventory.Inventory`                          90   <=====
:py:property:`sphobjinv.inventory.Inventory.count`                 90   
:py:method:`sphobjinv.inventory.Inventory.data_file`               90   
:py:attribute:`sphobjinv.inventory.Inventory.header_preamble`      90   
:py:attribute:`sphobjinv.inventory.Inventory.header_project`       90   
:py:attribute:`sphobjinv.inventory.Inventory.header_version`       90   
:py:attribute:`sphobjinv.inventory.Inventory.header_zlib`          90   
:py:method:`sphobjinv.inventory.Inventory.json_dict`               90   
:py:attribute:`sphobjinv.inventory.Inventory.objects`              90   
:py:property:`sphobjinv.inventory.Inventory.objects_rst`           90   
:py:attribute:`sphobjinv.inventory.Inventory.project`              90   
:py:attribute:`sphobjinv.inventory.Inventory.source_type`          90   
:py:method:`sphobjinv.inventory.Inventory.suggest`                 90   
:py:attribute:`sphobjinv.inventory.Inventory.version`              90   
:std:doc:`api/inventory`                                           90   
:py:exception:`sphobjinv.error.SphobjinvError`                     63   
 
```

For this docset, I don't have it displaying the fully specified object names (`add_module_names = False` in `conf.py`), but when I switch that to `True`, the copy of `Inventory`'s docs that comes from `.. automodule:: sphobjinv.inventory :members:` shows up like this:

![image](https://user-images.githubusercontent.com/11325439/163094290-31d4b1a0-3432-45a8-a32b-812814b84c99.png)

BUT, the copy of `Inventory`'s docs that comes from `.. autoclass:: sphobjinv.Inventory` shows up like this:

![image](https://user-images.githubusercontent.com/11325439/163094357-03104f39-cae0-42fd-9403-ab1e29b4be39.png)

Seems like what you want?

**Comment by asmeurer** [comment]
I'll have to play around with that. I suspect one problem will be our use of `automodule`, combined with some names being in `__init__.py` and some not. But between this and some other issues, I'm starting to think we may have to either abandon automodule (and make extensive use of external coverage tests to make sure we don't miss anything) or else somehow hook into it to modify what it does.

**Comment by asmeurer** [comment]
Actually rereading the OP of this issue I did touch on this a little bit. I might try a little more experimentation with it later, though.

**Comment by bskinn** [comment]
> I suspect one problem will be our use of `automodule,`

What is `automodule` actually doing for you? If the page for `sympy.sets.sets` is representative of the broader docs, I think all `automodule` is doing is adding an entry for the module to the index, and not adding any actual content to the doc page. It *might* be setting the subpackage reference from which the other autodoc directives are imported, but I'm not 100% sure on that.

Since you aren't using `:members:` with `automodule`, and the remainder of the page content (all of the `.. autoclass::` directives, &amp;c.) is not contained within the `.. automodule::` suite, I think `.. module::` would suffice just as well. Depends on whether that 'setting the subpackage reference' behavior' is happening or not.

Are you using `:members:` with `automodule`, for some modules? 

> combined with some names being in `__init__.py` and some not.

Does this mean that some things are meant to be imported from shallower in the package hierarchy from where they're defined (e.g., `sympy.Set`), but others are meant to be imported closer to, or at, the point in the package hierarchy  where they're defined?

----

Are you aware of the `:exclude-members:` option to `autodoc`? From [here](https://www.sphinx-doc.org/en/master/usage/extensions/autodoc.html):

> * The directives supporting member documentation also have a `exclude-members` option that can be used to exclude single member names from documentation, if all members are to be documented.
> 
> *New in version 0.6.*

I'm picturing three scenarios for a given module:

1. A module `sympy.foo.bar` contains *only* objects that are meant to be imported as `from sympy.foo.bar import baz`.
    - In this case, `.. automodule:: sympy.foo.bar  :members:` should satisfy.
2. A module `sympy.foo.bar` contains a *mix* of objects that are meant to be imported as `from sympy.foo.bar import baz` and as (e.g.) `from sympy import quux`.
    - In this case, `.. automodule:: sympy.foo.bar  :members: :exclude-members: quux`, plus explicit `.. autoclass:: sympy.quux` entries for each `quux`-like, should satisfy.
        - The main friction here is that all `quux`-likes would appear only at the top or bottom of the doc page (depending on whether they're defined inside or outside of the `.. automodule::` suite).
3. A module `sympy.foo.bar` has *only* objects that are meant to be imported as `from sympy import quux`.
    - In this case, `.. module:: sympy.foo.bar` combined with individual (e.g.) `.. autoclass:: sympy.quux` entries should suffice.

(3) is the case for [`sympy.sets.sets`](https://raw.githubusercontent.com/sympy/sympy/master/doc/src/modules/sets.rst), as far as I can tell, and you already have that same coverage problem with it, regardless of whether your `.. autoclass::` directives point to `Set` or `sympy.Set`, so you wouldn't lose much by switching it to `sympy.Set`.

----

High level, IMO, the API doc coverage question is more or less orthogonal to the 'show the right import location for `__init__.py`-ed objects' question.

Automatically finding the highest level in the package hierarchy where each object is importable (if this is even the right logic!) feels like it would be a major PITA to me. But, the docs are large enough that the investment may be worth it.

Separately, though, any use of `module`/`automodule` without `:members:` is going to introduce the possibility of a coverage problem. Trying to solve the two problems together seems like a nightmare to me...but, again, the docset is large enough that it may be the lesser of two monsters.

(Sorry if this turned into a long exercise in circling around and back to the status quo conundrum... 😬)

**Comment by ahonorat** [comment]
Hello,
I'm starting to use sympy in a project (not public yet), and I think I got a related problem while generating the documentation of my project with pydoctor: pydoctor cannot link unqualified names to the fully qualified sympy objects and encounters some errors while parsing the inventory file.

Any clue about how to avoid this?  
I'm using sympy 1.14 and pydoctor 25.4, I did not tried with other documentation tools.

Here is a MWE:
```python
#! /usr/bin/env python

import sympy as sy

def get_sy_expr_x1() -> sy.Expr:
    """Create the vector based symbol x and return index 1.

    Returns:
        sy.Expr: The symbolic expression x[1].
    """
    sym = sy.IndexedBase("x")
    return sym[1]

if __name__ == "__main__":
    print(get_sy_expr_x1())
```

Failing call to pydoctor:
```bash
pydoctor --docformat=google --intersphinx=https://docs.sympy.org/latest/objects.inv sympy_doc_bug.py
```
Outputs:

> Failed to parse line "D arm." for https://docs.sympy.org/latest
> Failed to parse line "series element representing the tendon, CE is the contractile" for https://docs.sympy.org/latest
> Failed to parse line "element, EE is the parallel element representing the elasticity of" for https://docs.sympy.org/latest
> Failed to parse line "the muscle fibers, and DE is the damper." for https://docs.sympy.org/latest
> adding module /tmp/sympy_doc_bug.py
> 1 modules and packages discovered
> Guessing 'sympy_doc_bug' for project name.
> 1/1 modules processed, 4 warnings
> writing html to apidocs using pydoctor.templatewriter.writer.TemplateWriter
> starting ModuleIndexPage ...took 0.003491s
> starting ClassIndexPage ...took 0.001017s
> starting NameIndexPage ...took 0.001190s
> starting UndocumentedSummaryPage ...took 0.000917s
> starting HelpPage ...took 0.029364s
> starting AllDocuments ...took 0.006295s
> starting lunr search index ...took 0.001028s
> 1/1 pages written
> /tmp/sympy_doc_bug.py:12: Cannot find link target for "sympy.Expr", resolved from "sy.Expr" (you can link to external docs with --intersphinx)
> Generating objects inventory at apidocs/objects.inv

**Comment by oscargus** [comment]
Although it is importable as sympy.Expr, it really resides as sympy.core.expr.Expr

One would maybe expect pydoctor to figure that out, but the best bet is to use the fully specified path.

**Comment by asmeurer** [comment]
I think that's effectively the same problem we're discussing here. You'll most likely just have to use the fully qualified name in your docs. You can use the `~.` shorthand to avoid it being spelled out in the rendered docs (I'm not completely sure how that looks for external references so you may need to play around with it). 

Maybe it is possible for us to fix this in the inventory file, but my guess that whatever we'd need to do to fix that is also what we'd need to do to make general non-fully qualified references work in our own docs as well (which is something I was never able to figure out, but we definitely would like to be able to do).

**Comment by bskinn** [comment]
I believe `~` does the same thing for internal and external references: only renders the name of the entity itself. So, ```:class:`~sympy.sets.sets.Set` ``` would render as just `Set`, and ```:class:`~pathlib.Path` ``` would render as just `Path`.

**Comment by ahonorat** [comment]
Thanks for your replies!

> Although it is importable as sympy.Expr, it really resides as sympy.core.expr.Expr
> 
> One would maybe expect pydoctor to figure that out, but the best bet is to use the fully specified path.

Well it seems that pydoctor does not inspect the imported libraries :/
I guess you are right about using the fully specified path until either you manage to have non-fully qualified references in the SymPy doc or I tweak pydoctor somehow ... 

About the other issue ("Failed to parse line ..."), I dug into the objets.inv with both sphobjinv and pydoctor
and I think I understand why. I will open another issue for that point, it should be easy to fix.

**Comment by tristanlatr** [comment]
Hello, pydoctor maintainer here.

 I don’t think that other documentation tools like Sphinx do any kind of supplementary magic to map implementation location to publicly importable location of external libraries. Especially when the only thing the tool knows about the library is its inventory file (the library itself does not have to be installed for pydoctor to generate the documentation for the depending packages - even the depending package does not have to be installed since it operates on the AST solely). 

Now, if you find out that Sphinx or other documentation tools does do some extra magic to circumvent that issue, I would be happy to look more closely to the implementation of such logic and try to have something similar in pydoctor. 

Many thanks,
